compiler-plugin/src/main/java/io/ballerina/lib/data/csvdata/compiler/CsvDataTypeValidator.java

```diff
@@ -0,0 +1,644 @@
+        }
+    }
+
+    private void validateTupleType(Location currentLocation, SyntaxNodeAnalysisContext ctx) {
+        // Currently, this is unsupported.
+        reportDiagnosticInfo(ctx, Optional.ofNullable(currentLocation), CsvDataDiagnosticCodes.UNSUPPORTED_TYPE);
+    }
+
+    private void validateArrayType(ArrayTypeSymbol typeSymbol,
+                                   Location currentLocation, SyntaxNodeAnalysisContext ctx) {
+        if (!isSupportedArrayMemberType(ctx, currentLocation, typeSymbol.memberTypeDescriptor())) {
+            reportDiagnosticInfo(ctx, Optional.ofNullable(currentLocation), CsvDataDiagnosticCodes.UNSUPPORTED_TYPE);
+        }
+    }
+
+    private void validateUnionType(UnionTypeSymbol unionTypeSymbol,
+                                   Location currentLocation, SyntaxNodeAnalysisContext ctx) {
+        List<TypeSymbol> memberTypeSymbols = unionTypeSymbol.memberTypeDescriptors();
+        for (TypeSymbol memberTypeSymbol : memberTypeSymbols) {
+            validateExpectedType(memberTypeSymbol, currentLocation, ctx);
+        }
+    }
+
+    private boolean isSupportedArrayMemberType(SyntaxNodeAnalysisContext ctx,
+                                               Location currentLocation, TypeSymbol typeSymbol) {
+        if (typeSymbol.typeKind() == TypeDescKind.TYPE_REFERENCE) {
+            typeSymbol = ((TypeReferenceTypeSymbol) typeSymbol).typeDescriptor();
+        }
+        if (typeSymbol.typeKind() == TypeDescKind.INTERSECTION) {
+            typeSymbol = getRawType(typeSymbol);
+        }
+        TypeDescKind kind = typeSymbol.typeKind();
+        if (kind == TypeDescKind.TYPE_REFERENCE) {
+            kind = ((TypeReferenceTypeSymbol) typeSymbol).typeDescriptor().typeKind();
+        }
+
+        switch (kind) {
+            case ARRAY, MAP, UNION, INTERSECTION -> {
+                return true;
+            }
+            case RECORD -> validateRecordFields(ctx, currentLocation, typeSymbol);
+            case TUPLE -> validateTupleMembers(ctx, currentLocation, typeSymbol);
+            default -> {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private void validateTupleMembers(SyntaxNodeAnalysisContext ctx, Location currentLocation, TypeSymbol typeSymbol) {
+        if (typeSymbol.typeKind() == TypeDescKind.TYPE_REFERENCE) {
+            typeSymbol = ((TypeReferenceTypeSymbol) typeSymbol).typeDescriptor();
+        }
+        TupleTypeSymbol tupleTypeSymbol = (TupleTypeSymbol) typeSymbol;
+        tupleTypeSymbol.memberTypeDescriptors().forEach(symbol ->
+                validateNestedTypeSymbols(ctx, currentLocation, symbol, false));
+        Optional<TypeSymbol> restSymbol = tupleTypeSymbol.restTypeDescriptor();
+        if (restSymbol.isPresent()) {
+            TypeSymbol restSym = restSymbol.get();
+            validateNestedTypeSymbols(ctx, currentLocation, restSym, false);
+        }
```

Comment:
I think we can do this cleaner as well
```suggestion
        tupleTypeSymbol.restTypeDescriptor().ifPresent(restSym -> validateNestedTypeSymbols(ctx, currentLocation, restSym, false));
```